tests/main/snap-user-service-upgrade-failure/task.yaml

```diff
@@ -0,0 +1,65 @@
+summary: On upgrade failures, user session daemons are restarted
+
+systems:
+    # Ubuntu 14.04's systemd doesn't have user@.service
+    - -ubuntu-14.04-*
+    # Amazon Linux 2 gives error "Unit user@12345.service not loaded."
+    - -amazon-linux-2-*
+    # Centos 7 gives error "Unit user@12345.service not loaded."
+    - -centos-7-*
+
+kill-timeout: 5m
+
+prepare: |
+    # Ensure that snapd.session-agent.socket is enabled.  This may not
+    # be the case on distributions where presets have been used to
+    # disable it.
+    if [ ! -L /usr/lib/systemd/user/sockets.target.wants/snapd.session-agent.socket ] &&
+            ! systemctl --user --global is-enabled snapd.session-agent.socket; then
+        systemctl --user --global enable snapd.session-agent.socket
+        touch agent-was-enabled
+    fi
+    snap set system experimental.user-daemons=true
+    session-tool --kill-leaked
+    session-tool -u test --prepare
+    session-tool -u test systemctl --user start sockets.target
+
+restore: |
+    session-tool -u test --restore
+    #shellcheck source=tests/lib/user.sh
+    . "$TESTSLIB/user.sh"
+    purge_user_session_data
```

Comment:
Is this still required? What is being purged here that session shutdown is not removing?

Reply:
This was an attempt to fix the failure in this run:

https://github.com/snapcore/snapd/commit/6144f2f89eb8cac982159b31fa04f02b49dab249/checks/642364308/logs

The `tests/main/user-mounts` test started failing when I migrated my spread tests over to session-tool, with it being clear that user `test`'s XDG_RUNTIME_DIR still existed and contained a mount of xdg-document-portal, causing it's prepare step to fail (where it tries to remove everything in XDG_RUNTIME_DIR).

Adding the purge_user_session_data calls led to different failures:

https://github.com/snapcore/snapd/commit/3ded2e359f8462a26efc4a622f22dfbf69d2b8da/checks/642749712/logs

There are now failures in session-tool based tests like:

```
2020-05-04T14:06:18.0486363Z + session-tool --kill-leaked
2020-05-04T14:06:18.0486620Z + session-tool -u test --prepare
2020-05-04T14:06:18.0486913Z + session-tool -u test systemctl --user start sockets.target
2020-05-04T14:06:18.0487035Z Failed to connect to bus: No such file or directory
```

This would seem to indicate that the user session survived a previous test, but my purge_user_session_data call deleted the session bus socket causing the test to fail.  That would seem to indicate that the session is still active after `session-tool -u test --restore` returns.